File: 이상욱/프로그래머스/m4w13/PG92334.java
package m4w13;

import java.util.*;


public class PG92334 {
	
	
}
class Solution {
    public int[] solution(String[] id_list, String[] report, int k) {
        int[] answer = new int[id_list.length];
        
        Arrays.sort(report);
        
        HashMap<String, Integer> hm = new HashMap<>();
        HashMap<String, ArrayList<String>> hm1 = new HashMap<>();
        
        for(int i = 0; i < id_list.length; i++) {
        	hm.put(id_list[i], 0);
        	hm1.put(id_list[i], new ArrayList<String>());
        }
        
        
        String preReport = "";
        
        for(int i = 0; i < report.length; i++) {
        	if(preReport.equals(report[i])) {
        		continue;
        	}
        	
        	preReport = report[i];
        	
        	String[] content = report[i].split(" ");
        	
        	String p1 = content[0];
        	String p2 = content[1];
        	
        	hm.put(p2, hm.getOrDefault(p2, 0)+1);
        	
        	ArrayList<String> list = hm1.getOrDefault(p1, new ArrayList<String>());
        	list.add(p2);
        	hm1.put(p1, list);
        	
        }
        
        for(String key : hm.keySet()) {
        	if(hm.get(key) >= k) {
        		for(int i = 0; i < id_list.length; i++) {
        			String p = id_list[i];
        			if(hm1.get(p).contains(key)) {
        				answer[i]++;
        			}
        		}
        	}
        }
        
        
        return answer;
    }
}

File: 이상욱/프로그래머스/m4w13/PG92334.kt
package m4w13

class Solution {
    fun solution(id_list: Array<String>, report: Array<String>, k: Int): IntArray {
        val answer = IntArray(id_list.size)

        report.sort()

        val hm = hashMapOf<String, Int>()
        val hm1 = hashMapOf<String, ArrayList<String>>()

        id_list.forEach {
            hm[it] = 0
            hm1[it] = arrayListOf()
        }

        var preReport = ""
        for (i in report.indices) {

            if (preReport == report[i]) {
                continue
            }
            preReport = report[i]

            val content = report[i].split(" ")
            val p1 = content[0]
            val p2 = content[1]

            hm[p2] = hm[p2]!! + 1
            hm1[p1]!!.add(p2)

        }

        for (i in hm) {
            if (i.value >= k) {
                id_list.forEachIndexed { index, s ->
                    val p = s
                    if (hm1[p]!!.contains(i.key)) {
                        answer[index]++
                    }
                }
            }
        }


        return answer
    }
}